shell/platform/windows/flutter_windows_engine.cc

```diff
@@ -281,6 +287,18 @@ bool FlutterWindowsEngine::RunWithEntrypoint(const char* entrypoint) {
   return true;
 }
 
+void FlutterWindowsEngine::OnVsync(intptr_t baton) {
```

Comment:
There is an inaccuracy here that will make the embedder pump frames as fast possible with no regard for display capabilities. Here is why:

* `frame_start_time_nanos` as [documented](https://github.com/flutter/engine/blob/c120a50ecb3b32927a387900418a6d09dfe1e287/shell/platform/embedder/embedder.h#L1964) says "The point at which the vsync event occurred or will occur". By saying that the vsync event has already occurred, the engine will begin processing the frame immediately. It won't be aligned to any frame boundary.
* `frame_target_time_nanos` as [documented](https://github.com/flutter/engine/blob/c120a50ecb3b32927a387900418a6d09dfe1e287/shell/platform/embedder/embedder.h#L1969) is only a hint to the engine as to when the next vsync is expected to occur. This allows it to more effectively schedule new-space GCs.

As written, the engine will ask the embedder for the next vsync, the embedder will tell the engine that one has just happened, the engine will process the frame and (potentially immediately) ask for the next vsync which the embedder will also keep saying has already happened. This will continue till the GPU is overwhelmed and the pipelining mechanism kicks in to make the engine back off.

To implement `FlutterEngineOnVsync` you need two pieces of information:
1. A signal from the embedder that a vsync occurred.
2. A indication from the embedder about what the delta between vsyncs is expected to be.

In this patch you have (2) but not (1). But, that not the end of the world. You aren't trying to schedule frames with actual vsync, just attempting to support variable refresh rates. So, you can fudge the information you don't have yet for (1). You can snap `frame_start_time_nanos` to the next timepoint that is wholly divisible by the frame interval you have (2). We do this in [the fallback waiter](https://github.com/flutter/engine/blob/d64eabe4d3a578a6976d188ca8013144fc47fb13/shell/common/vsync_waiter_fallback.cc#L13) if you want to see this technique. The fallback waiter doesn't have (1) just like you and it assumes (2) to be 1/60 seconds.

Ideally, you will have both pieces of information but I am not sure how difficult that is to achieve on Windows.

Also, I urge you to capture a trace on the device after you wire this up. Any frame scheduling issues become more immediately apparent when viewed this way. In fact, I would appreciate a trace after attached to the review just in case there is something I missed.

Reply:
Thank you @chinmaygarde , this is a ton of help 😄

Reply:
I spent some more time digging into this and found that 2) alone is not sufficient to consistently hit 144 fps on my local set up.

Thankfully I found that I can actually grab the time of the last vsync from the DWM. The combination of that, the framerate, and some QPC logic has been enough for me to get consistent 144 fps. I don't believe this has the same issue, but I need to do some more local testing, cover some of the edge cases if either DWM or QPC aren't supported, and get to working on unit tests.

Reply:
> 2) alone is not sufficient to consistently hit 144 fps on my local set up.

Is this because the scene is too complicated or because of scheduling?